astropy/units/equivalencies.py

```diff
@@ -744,24 +744,32 @@ def assert_is_spectral_unit(value):
 
 def pixel_scale(pixscale):
     """
-    Convert between pixel distances (in units of ``pix``) and angular units,
+    Convert between pixel distances (in units of ``pix``) and other units,
     given a particular ``pixscale``.
 
     Parameters
     ----------
     pixscale : `~astropy.units.Quantity`
-        The pixel scale either in units of angle/pixel or pixel/angle.
+        The pixel scale either in units of <unit>/pixel or pixel/<unit>.
     """
-    if pixscale.unit.is_equivalent(si.arcsec/astrophys.pix):
-        pixscale_val = pixscale.to_value(si.radian/astrophys.pix)
-    elif pixscale.unit.is_equivalent(astrophys.pix/si.arcsec):
-        pixscale_val = (1/pixscale).to_value(si.radian/astrophys.pix)
-    else:
-        raise UnitsError("The pixel scale must be in angle/pixel or "
-                         "pixel/angle")
 
-    return Equivalency([(astrophys.pix, si.radian,
-                         lambda px: px*pixscale_val, lambda rad: rad/pixscale_val)],
+    # Try inspect the equivalent unit.
+    _scale = pixscale.decompose()
+    _dims = dict(zip(_scale.unit.bases, _scale.unit.powers))
+    pix_power = _dims.get(astrophys.pix, 0)
+    if pix_power == -1:
+        value_per_pix = _scale
+    elif pix_power == 1:
+        value_per_pix = 1. / _scale
+    else:
+        raise UnitsError(
+                "The pixel scale unit must have"
+                " pixel dimensionality of 1 or -1.")
+    u_phys = value_per_pix.unit * astrophys.pix
+
+    return Equivalency([(astrophys.pix, u_phys,
+                         lambda px: px * value_per_pix,
```

Comment:
Multiplication and division are the defaults for equivalencies, so in principle there should be no need to define these `lambda` expressions. However, that relies on getting the replacement unit - here `u_phys` - right. The way to do that is to ensure `u_phys` also contains the value. To get that, one can replace as follows (where I also adjusted according to the comment above):
```
    pixscale_unit = Unit(pixscale)
    decomposed = pixscale_unit.decompose()
    dimensions = dict(zip(decomposed.bases, decomposed.powers))
    pix_power = dimensions.get(astrophys.pix, 0)
    if pix_power == -1:
        physical_unit = pixscale_unit * astrophys.pix
    elif pix_power == 1:
        physical_unit = astrophys.pix / pixscale_unit
    else:
        raise UnitsError(
                "The pixel scale unit must have"
                " pixel dimensionality of 1 or -1.")

    return Equivalency([(astrophys.pix, physical_unit)],
                       "pixel_scale", {'pixscale': pixscale})
```

Reply:
Thanks for the suggestion. I made the changes accordingly, and the code now become quite neat. I did some reordering of the variable defs to make it more readable.